inst/odin/compare_carehomes.cpp

```diff
@@ -0,0 +1,215 @@
+#include <dust/densities.hpp>
+
+template <typename real_t>
+real_t ll_nbinom(real_t data, real_t model, real_t kappa, real_t exp_noise,
+                 dust::rng_state_t<real_t>& rng_state) {
+  if (std::isnan(data)) {
+    return 0;
+  }
+  real_t mu = model + dust::distr::rexp(rng_state, exp_noise);
+  return dust::dnbinom(data, kappa, mu, true);
+}
+
+template <typename real_t>
+real_t ll_binom(real_t data_x, real_t data_size, real_t model_prob) {
+  if (std::isnan(data_x) || std::isnan(data_size)) {
+    return 0;
+  }
+  return dust::dbinom(data_x, data_size, model_prob, true);
+}
+
+template <typename real_t>
+real_t ll_betabinom(real_t data_x, real_t data_size, real_t model_prob,
+                    real_t rho) {
+  if (std::isnan(data_x) || std::isnan(data_size)) {
+    return 0;
+  }
+  return dust::dbetabinom(data_x, data_size, model_prob, rho, true);
+}
+
+template <typename real_t>
+real_t test_prob_pos(real_t pos, real_t neg, real_t sensitivity,
+                     real_t specificity, real_t exp_noise,
+                     dust::rng_state_t<real_t>& rng_state) {
+  // We add some exponential noise to the number of positives and negatives
+  // to help ensure prob_pos is not 0 or 1. If e.g. prob_pos were 0 and there
+  // were individuals who tested positive, this would result in a weight of 0
+  // for a particle. If all particles have weights of 0, the particle filter
+  // breaks. The exponential noise produces small non-zero weights in these
+  // circumstances to prevent the particle filter from breaking.
+  pos += dust::distr::rexp(rng_state, exp_noise);
+  neg += dust::distr::rexp(rng_state, exp_noise);
+  return (sensitivity * pos + (1 - specificity) * neg) / (pos + neg);
+}
+
+// Lots of data items here. Even though some of these are counts (in
+// fact, most of them are counts) we'll pull them in as doubles
+// because that will help with the NA handling. Once this is working
+// reliably we could cast these explicitly as ints, but there is not
+// likely to be any major gain here.
+//
+// [[odin.dust::compare_data(icu = double)]]
```

Comment:
Is there any way to have these as real_t, or turned into the right type by odin.dust?

Reply:
Good idea, though it will require odin.dust support: added as https://github.com/mrc-ide/odin.dust/issues/57